include/gc/Dialect/Microkernel/MicrokernelOps.td

```diff
@@ -10,5 +10,100 @@
 #define MICROKERNEL_OPS
 
 include "MicrokernelDialect.td"
+include "gc/Dialect/Microkernel/MicrokernelEnum.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
 
-#endif // MICROKERNEL_OPS
\ No newline at end of file
+class StaticMemRefRankOf<list<Type> allowedTypes, list<int> ranks> :
+     Type<And<[MemRefOf<allowedTypes>.predicate,
+              HasAnyRankOfPred<ranks>, HasStaticShapePred]>, 
+              !interleave(!foreach(rank, ranks, rank # "D"), "/") # " static " # 
+              MemRefOf<allowedTypes>.summary, "::mlir::MemRefType">;
+
+def Microkernel_BrgemmDispatchOp : Microkernel_Op<"brgemm.dispatch", [Pure]> {
+  let summary = "JIT the brgemm microkernel given the parameters";
+  let description = [{
+    The operation has the following arguments: 1) m, n, k, lda, ldb, ldc, stride_a and stride_b. 
+    Inputs is a dense attribute of I64 elements. 2) flags carry information on
+    the different flags that can be used for brgemm like whether beta == 0 or strided batch. For
+    more details, see: `Microkernel_BrgemmFlags`. 3) data_types of operand A & B.
+    Outpus is the id of JITed kernel.
+  }];
+
+  let arguments = (ins
+    ConfinedAttr<DenseI64ArrayAttr,
+                [DenseArrayNonNegative<DenseI64ArrayAttr>]>:$inputs,
+    TypedArrayAttrBase<Microkernel_BrgemmFlags, "brgemm flags">:$flags,
+    TypedArrayAttrBase<TypeAttr, "brgemm dtypes">:$data_type);
+
+  let results = (outs I64:$results);
+  let hasCustomAssemblyFormat = 1;
+  let hasVerifier = 1;
+}
+
+def Microkernel_BrgemmPrologueOp : Microkernel_Op<"brgemm.prologue"> {
+  let summary = "Prologue before executing the JITed brgemm "
+                "microkernel, and the context is considered core-level";
+  let description = [{
+    The operation has the following arguments: Input is the id of JITed kernel.
+    There is no output.
+  }];
+
+  let arguments = (ins I64:$inputs);
+
+  let assemblyFormat = [{
+    `(` $inputs `)`
+    attr-dict `:` functional-type($inputs, results)
+  }];
+}
+
+def Microkernel_BrgemmEpilogueOp : Microkernel_Op<"brgemm.epilogue"> {
+  let summary = "Epilogue after executing the JITed brgemm microkernel";
+  let description = [{
+    The operation has the following arguments: Input is the id of JITed kernel.
+    There is no output.
+  }];
+
+  let arguments = (ins I64:$inputs);
+
+  let assemblyFormat = [{
+    `(` $inputs `)`
+    attr-dict `:` functional-type($inputs, results)
+  }];
+}
+
+def BrgemmMemRef : AnyTypeOf<[StaticMemRefRankOf<[F32, BF16, SI32, SI8, UI8], [2, 3, 4]>, I64]>;
```

Comment:
Could you please articulate somewhere here what the exact restrictions on the input type are?

Reply:
Changed the type name to `BrgemmMemRefOrI64` to better reflect its restrictions.
Also added comments here to explain the exact restrictions for BrgemmMemRef for each operand.